File: src/test/java/io/protobase/f7/formulas/math/SINHTest.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.errors.NAException;
import io.protobase.f7.errors.ValueException;
import io.protobase.f7.models.Grid;
import io.protobase.f7.testutils.TestFormula;
import org.junit.Test;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

public class SINHTest extends TestFormula {
  @Test
  public void testApply() {
    assertThat(SINH.SELF.apply(null, 10.0)).isEqualTo(Math.sinh(10.0));
    assertThat(SINH.SELF.apply(null, 1.2)).isEqualTo(Math.sinh(1.2));
    assertThat(SINH.SELF.apply(null, 11.11)).isEqualTo(Math.sinh(11.11));
    assertThat(SINH.SELF.apply(null, 0.0)).isEqualTo(Math.sinh(0.0));
    assertThat(SINH.SELF.apply(null, 8881.0)).isEqualTo(Math.sinh(8881.0));
    assertThat(SINH.SELF.apply(null, 2.0)).isEqualTo(Math.sinh(2.0));
    assertThat(SINH.SELF.apply(null, 4.0)).isEqualTo(Math.sinh(4.0));
    assertThat(SINH.SELF.apply(null, -4.0)).isEqualTo(Math.sinh(-4.0));
    assertThat(SINH.SELF.apply(null, -10124.0)).isEqualTo(Math.sinh(-10124.0));
  }

  @Test
  public void testApply_stringConversion() {
    assertThat(SINH.SELF.apply(null, "10.0")).isEqualTo(Math.sinh(10.0));
  }

  @Test
  public void testApply_grid() {
    assertThat(SINH.SELF.apply(null,
        Grid.builder().add(0, 0, -4.0).add(0, 1, "Don't mind me.").build()
    )).isEqualTo(Math.sinh(-4.0));
  }

  @Test
  public void test_lookup() {
    SINH F = new SINH(lookup, collateralLookup);
    when(collateralLookup.apply(A1, M22_RANGE)).thenReturn(10.0);
    assertThat(F.apply(A1, M22_RANGE)).isEqualTo(Math.sinh(10.0));
    verify(collateralLookup).apply(A1, M22_RANGE);
    verifyNoMoreInteractions(collateralLookup);
  }

  @Test
  public void testApply_errorsPassThrough() {
    assertThat(SINH.SELF.apply(null, new ValueException())).isEqualTo(new ValueException());
  }

  @Test
  public void testApply_argumentsMismatch() {
    assertThat(SINH.SELF.apply(null, "A", "Too many")).isEqualTo(new NAException());
  }
}

File: src/test/js/formulas/math/SINHTest.ts
import { assert } from "chai";
import { it, describe } from "../../testutils/TestUtils";
import { stub } from "sinon";
import { F7ExceptionName } from "../../../../main/js/errors/F7ExceptionName";
import { NAException } from "../../../../main/js/errors/NAException";
import { ValueException } from "../../../../main/js/errors/ValueException";
import { SINH } from "../../../../main/js/formulas/math/SINH";
import { Grid } from "../../../../main/js/models/common/Grid";
import { CollateralLookupFunction, LookupFunction } from "../../../../main/js/models/common/Types";
import { CommonModels } from "../../testutils/CommonModels";

describe("SINH", function () {
  it("should work with valid numbers", function () {
    assert.deepEqual(SINH.SELF.run(null, 10), Math["sinh"](10));
    assert.deepEqual(SINH.SELF.run(null, 128731.2), Math["sinh"](128731.2));
    assert.deepEqual(SINH.SELF.run(null, 11.11), Math["sinh"](11.11));
    assert.deepEqual(SINH.SELF.run(null, 0), 0);
    assert.deepEqual(SINH.SELF.run(null, 88281), Math["sinh"](88281));
    assert.deepEqual(SINH.SELF.run(null, 2), Math["sinh"](2));
    assert.deepEqual(SINH.SELF.run(null, 4), Math["sinh"](4));
    assert.deepEqual(SINH.SELF.run(null, -4), Math["sinh"](-4));
    assert.deepEqual(SINH.SELF.run(null, -10124), Math["sinh"](-10124));
  });

  it("should work with strings", function () {
    assert.equal(SINH.SELF.run(null, "10"), Math["sinh"](10));
  });

  it("should do pass-through errors", function () {
    assert.deepEqual(SINH.SELF.run(null, new ValueException()), new ValueException());
  });

  it("should use lookup", function () {
    const lookup = stub();
    const collateralLookup = stub();
    const F = new SINH(lookup as LookupFunction, collateralLookup as CollateralLookupFunction);
    collateralLookup.withArgs(CommonModels.A1, CommonModels.M22_RANGE).returns(10);
    assert.deepEqual(F.run(CommonModels.A1, CommonModels.M22_RANGE), Math["sinh"](10));
    assert.isTrue(collateralLookup.calledWith(CommonModels.A1, CommonModels.M22_RANGE));
    assert.equal(collateralLookup.callCount, 1);
    assert.isTrue(lookup.notCalled);
  });

  it("should handle grids", function () {
    const one = Grid.builder().add(0, 0, 10).add(0, 1, "A").build();
    assert.deepEqual(SINH.SELF.run(null, one), Math["sinh"](10));
  });

  it("should return error when argument lengths are wrong", function () {
    assert.deepEqual((SINH.SELF.run(null) as NAException).name, F7ExceptionName.NA);
    assert.deepEqual((SINH.SELF.run(null, "A", "B") as NAException).name, F7ExceptionName.NA);
  });
});
